File: src/oasees_sdk/stack/demo-server/lib/ndm.js
// NDM - Node Data Manager
// Ensures data integrity via SHA-256 signing and publishes results

const EventEmitter = require('events');
const crypto = require('crypto');

class NDM extends EventEmitter {
  constructor(nds) {
    super();
    this.nds = nds || null;
    this.emit('action', { component: 'NDM', action: 'init', log: '[NDM] Initialized. Data integrity module ready.' });
  }

  signData(data) {
    const dataStr = JSON.stringify(data, Object.keys(data).sort());
    const signature = crypto.createHash('sha256').update(dataStr).digest('hex');
    this.emit('action', {
      component: 'NDM', action: 'signData',
      log: `[NDM] Data signed — SHA-256: ${signature.slice(0, 16)}...`,
      data: { inputSize: dataStr.length, signature }
    });
    return signature;
  }

  processOutput(data, jobId) {
    this.emit('action', {
      component: 'NDM', action: 'processOutput',
      log: `[NDM] Processing output data for integrity wrapping...`,
      data
    });

    // 1. Sign
    const signature = this.signData(data);

    // 2. Add metadata
    const payload = {
      data,
      signature,
      timestamp: Date.now() / 1000,
      integrity_proof: `proof-${signature.slice(0, 8)}`
    };

    // 3. Publish
    this.emit('action', {
      component: 'NDM', action: 'publish',
      log: `[NDM] Publishing signed result (M11) — integrity_proof: ${payload.integrity_proof}`,
      data: payload
    });

    if (this.nds && jobId) {
      this.nds.storeResult(jobId, payload);
    }

    return payload;
  }
}

module.exports = NDM;

File: src/oasees_sdk/stack/ndm.py

import hashlib
import json
import time

class NDM:
    """
    Node Data Manager (NDM) - Mock Implementation
    
    Responsible for:
    1. Ensuring data integrity (signing).
    2. Publishing results to the network/storage.
    """
    
    def __init__(self, nds=None):
        self.nds = nds
        print("[NDM] Initialized.")

    def sign_data(self, data):
        """
        Simulate data signing using a simple hash.
        """
        data_str = json.dumps(data, sort_keys=True)
        signature = hashlib.sha256(data_str.encode()).hexdigest()
        return signature

    def process_output(self, data, job_id=None):
        """
        Interface M11: Publish Signed Result (NGSI-LD Upsert).
        """
        print(f"[NDM] Processing output data: {data}")
        
        # 1. Sign
        signature = self.sign_data(data)
        
        # 2. Add metadata
        payload = {
            "data": data,
            "signature": signature,
            "timestamp": time.time(),
            "integrity_proof": f"proof-{signature[:8]}"
        }
        
        # 3. "Publish" to NDS or External Network
        print(f"[NDM] Publishing signed result: {json.dumps(payload)}")
        if self.nds and job_id:
            self.nds.store_result(job_id, payload)
        return payload
